keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,289 @@
+* updates the StorageVersion object, using the resourceVersion in the first step
+  to avoid conflicting with other API servers.
+* installs the resource handler
+
+We need to make sure that an API server establishes its membership in the list
+of participating API servers before reporting its preferred storage version via
+the StorageVersion API. Then the above procedure makes it impossible to mistake
+a new CandidateVersions entry as a stale one.
+
+### A nice property of the selected storage version
+
+A nice property of the selected storage version is that it is supported by all
+API servers during rolling upgrades/downgrades. This means that the proposed
+dynamic storage version mechanism does not introduce any API server downtime.
+
+The property is a by-product of the Kubernetes API [deprecation policy][] rule
+`#4a` and `#4b`. The following spreadsheet illustrates the fastest possible
+evolution of the storage version of a resource. It's easy to see in the
+spreadsheet that the "preferred storage version" of release *n* is part of the
+"supported versions" in releases *n±1*.
+
+|                           | release x | release x+1 | release x+2 | release x+3 |
+|---------------------------|-----------|-------------|-------------|-------------|
+| supported versions        | v1        | v1, v2      | v1, v2      | v2          |
+| preferred storage version | v1        | v1          | v2          | v2          |
+
+This fact, together with the following two facts, guarantee the "nice property".
+* if in an HA master, all API servers are of the same version, then the
+   preferred storage version is the same as the selected storage version;
+* during rolling upgrades/downgrades, the maximum version skew between any two
+   API servers is one.
+
+### Garbage collection of the StorageVersion objects of removed resources
+
+During bootstrap, the kube-apiserver
+* gets the list of all participating API servers,
+* gets all the StorageVersion objects,
+* removes the StorageVersion object if none of its CandidateVersions.APIServerID
+  is a participating API server.
+
+### Using the selected storage version when serializing data for etcd
+
+Depending on how strict the guarantee we want to provide, we have two
+alternative designs.
+
+1. strict guarantee: kube-apiserver always use the selected storage version
+
+To guarantee this, kube-apiserver needs to add a transaction condition to the
+write requests, to make sure the version it uses to serialize the API object is
+the selected storage version recorded in the API server.
+
+The performance cost is small, because all write operations from the
+kube-apiserver to etcd are transactional already.  For example, kube-apiserver
+checks if key does not already exist before committing a create operation. As
+another example, kube-apiserver checks that the resourceVersion is the latest
+before committing an update operation. Adding another transaction condition does
+not add extra RPC call.
```

Comment:
It would be good to run some benchmark of it to understand the impact of it. I hope it should be relatively small, but I don't know etcd internals well enough to know exactly what would happen.
@jpbetz - FYI